### PR TITLE
8261397: Try Catch Method Failing to Work When Dividing An Integer By 0

### PR DESCRIPTION
**Pull request by gerard-ziemski**
On Mac ARM hardware running x86 JDK under Rosetta emulation, a div by 0 instruction causes the VM to crash.

The proposed fix (a workaround) for hotspot is to add **FPE_FLTINV** to the signal handler.

The actual fix needs to be done in macOS by Apple as the expected signal type here is **FPE_FLTDIV**
This issue has been filed with Apple and they are tracking it.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261397](https://bugs.openjdk.java.net/browse/JDK-8261397): Try Catch Method Failing to Work When Dividing An Integer By 0

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to 0b8155117b88543637dbf4535351be8fb4afaeea
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to 8291664ed1f2e829d5e4bfa2186c6476ac49b9a0
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to 8291664ed1f2e829d5e4bfa2186c6476ac49b9a0
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2615/head:pull/2615`
`$ git checkout pull/2615`

### CODE REVIEW

**Comment by gerard-ziemski** [comment]
Simple native test case for those curious to try it out for themselves:

```c
/*
 Natively on x86_64 we get:
   sigHandler caught sig: 8, info->si_code: 7 [FPE_INTDIV]
 Under Rossetta for x86_64 binary we get:
   sigHandler caught sig: 8, info->si_code: 5 [FPE_FLTINV]
 */
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
static void sigHandler(int sig, siginfo_t* info, void* arg) {
  printf("sigHandler caught sig: %d, info->si_code: %d [", sig, info->si_code);
  switch (info->si_code) {
    case FPE_FLTDIV: printf("FPE_FLTDIV");break;
    case FPE_INTDIV: printf("FPE_INTDIV");break;
    case FPE_FLTINV: printf("FPE_FLTINV");break;
    default: printf("???");
  }
  printf("]\n");
  exit(sig);
}
int main(int argc, const char * argv[]) {
  struct sigaction sa;
  sigemptyset(&sa.sa_mask);
  sa.sa_sigaction = sigHandler;
  sa.sa_flags = SA_SIGINFO|SA_RESETHAND;
  sigaction(SIGFPE, &sa, NULL);
  volatile int i = 0;
  volatile int j = 47 / i;
  return j;
}
```

File: src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp [hunk 1]
```diff
@@ -472,7 +472,7 @@ bool PosixSignals::pd_hotspot_signal_handler(int sig, siginfo_t* info,
 
 #ifdef AMD64
       if (sig == SIGFPE  &&
-          (info->si_code == FPE_INTDIV || info->si_code == FPE_FLTDIV)) {
+          (info->si_code == FPE_INTDIV || info->si_code == FPE_FLTDIV MACOS_ONLY(|| info->si_code == FPE_FLTINV))) {
```
Comment (tstuefe) [inline src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp:475]: Can you add a comment stating that MacOs (or Rosetta?) misreports FLTDIV as FLTINV?

**Review by tstuefe — APPROVED** [review]
Seems reasonable. Remark below. Looks good otherwise.

Cheers, Thomas

File: src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp (hunk 1 above)
Comment (gerard-ziemski) [reply src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp:475]: Will do.

**Review by prrace — APPROVED** [review]
Looks reasonable to me. I suppose there is a BSD unix port that uses this code hence the MACOS_ONLY ?

**Comment by gerard-ziemski** [comment]
> Looks reasonable to me. I suppose there is a BSD unix port that uses this code hence the MACOS_ONLY ?

Exactly, any BSD platform uses this implementation. macOS is just one of them.

Thanks Phil!

File: src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/os_cpu/bsd_x86/os_bsd_x86.cpp:477]: nit - can you put a period at the end of L475.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. Nice job on keeping the work around very focused.

**Comment by gerard-ziemski** [comment]
> Thumbs up. Nice job on keeping the work around very focused.

Thank you Dan!

**Review by dholmes-ora — COMMENTED** [review]
Hi Gerard,

I'm concerned by the general problem of encountering new bugs because we are running on an x64 emulation mode on ARM! Is there any way to detect this Rosetta environment? Is there a way to know from the hs_err file (we don't have one for this issue yet) that we executed under Rosetta?

How will we track this bug on Apple's side, such that we can eventually remove this workaround?

The workaround itself seems okay as far as it goes but what happens if native code triggers a real FPE_FLTINV that we will now treat as "div by zero"?

Thanks,
David

**Comment by gerard-ziemski** [comment]
> Hi Gerard,
> 
> I'm concerned by the general problem of encountering new bugs because we are running on an x64 emulation mode on ARM! Is there any way to detect this Rosetta environment?

We can detect if we are running under emulation using this code provided by Apple [https://developer.apple.com/documentation/apple_silicon/about_the_rosetta_translation_environment](https://developer.apple.com/documentation/apple_silicon/about_the_rosetta_translation_environment?language=objc#3616845)

> Is there a way to know from the hs_err file (we don't have one for this issue yet) that we executed under Rosetta?

One can infer such scenario by looking at the log under the _system_ section:

`OS:uname: Darwin 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 x86_64`

We see **RELEASE_ARM64_T8101** type kernel running **x86_64** code, but we should detect this and report it to users directly. Filed a followup issue [JDK-8261966](https://bugs.openjdk.java.net/browse/JDK-8261966) to track it. Thank you for the bringing this point up!

> How will we track this bug on Apple's side, such that we can eventually remove this workaround?

I have filed the issue a while ago with Apple here [https://feedbackassistant.apple.com/feedback/8984455](https://feedbackassistant.apple.com/feedback/8984455), but please note that you probably need Mac developer account to be able to see it.

Apple has set the resolution of this issue as `Potential fix identified - For a future OS update`, but we will always need it for the already released macOS version.

> The workaround itself seems okay as far as it goes but what happens if native code triggers a real FPE_FLTINV that we will now treat as "div by zero"?

Excellent question. According to https://en.wikipedia.org/wiki/IEEE_754 invalid operation is `mathematically undefined, e.g., the square root of a negative number. By default, returns qNaN.` I will test this scenario out and report back...

Regardless though, we absolutely need this workaround for the time being, though we might want to use only in emulation environment.

**Comment by gerard-ziemski** [comment]
> > The workaround itself seems okay as far as it goes but what happens if native code triggers a real FPE_FLTINV that we will now treat as "div by zero"?
> 
> Excellent question. According to https://en.wikipedia.org/wiki/IEEE_754 invalid operation is `mathematically undefined, e.g., the square root of a negative number. By default, returns qNaN.` I will test this scenario out and report back...

According the **man pages** for `log()` function:

`log(x), log2(x), and log10(x) return a NaN and raise the "invalid" floating-point exception for x < 0.`

and the **man pages** for `sqrt()` function:

`sqrt(x) returns a NaN and generates a domain error for x < 0`

but what I see in our signal handler is a signal of type `FPE_FLTDIV` returned **not** `FPE_FLTINV` for `sqrt(-1.0)` and `log(-1.0)`

I can't seem to be able to get `FPE_FLTINV` at all - it's seems that it's unused natively on `macOS x86_64` so we are good I think.

**Comment by gerard-ziemski** [comment]
> > I can't seem to be able to get `FPE_FLTINV` at all - it's seems that it's unused natively on `macOS x86_64` so we are good I think.
> 
> Ok, but I think we should fix JDK-8261966 and include
> VM_Version::is_cpu_emulated() so that we use it to gate the workaround
> at runtime. That at least minimises the exposure to the environments
> that actually need the workaround.

I added the check in this fix - I don't think we should get this issue blocked by JDK-8261966

For JDK-8261966 we can further optimize the check to cache the results, if needed, and we can have a non rushed discussion on how exactly we want the API named (ex. I would prefer to follow Apple's usage and name it is_process_translated()) and how exactly we want it to look in the hs_err log files.


cheers

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Gerard,

The new VM_Version check should be in vm_version_bsd_x86.cpp so that we don't need to ifdef the code in the generic x86 file.

I don't think we need make is_cpu_emulated available on all platforms, jusyt bsd-x86.

Thanks,
David

**Comment by gerard-ziemski** [comment]
> Hi Gerard,
> 
> The new VM_Version check should be in vm_version_bsd_x86.cpp so that we don't need to ifdef the code in the generic x86 file.
> 
> I don't think we need make is_cpu_emulated available on all platforms, jusyt bsd-x86.
> 
> Thanks,
> David

I figured we would need eventually something similar for other OS (any x86_64 based OS being ported to aarch64), so I put the API into `vm_version_x86`.

`os_bsd_x86.hpp/.cpp` still needs the `#if __APPLE__`, but I did remove the API from ` vm_version_x86`

**Comment by gerard-ziemski** [comment]
> Hi Gerard,
> 
> The new VM_Version check should be in vm_version_bsd_x86.cpp so that we don't need to ifdef the code in the generic x86 file.
> 
> I don't think we need make is_cpu_emulated available on all platforms, jusyt bsd-x86.
> 
> Thanks,
> David

I did not realize that there was `vm_version_bsd_x86.cpp` available till you pointed it out.

Fixed.

**Comment by gerard-ziemski** [comment]
> > Hi Gerard,
> > The new VM_Version check should be in vm_version_bsd_x86.cpp so that we don't need to ifdef the code in the generic x86 file.
> > I don't think we need make is_cpu_emulated available on all platforms, jusyt bsd-x86.
> > Thanks,
> > David
> 
> I did not realize that there was `vm_version_bsd_x86.cpp` available till you pointed it out.
> 
> Fixed.

Hmm, I still think we need `#ifdef __APPLE__` here as well, there could be other BSD x86 OSes...

**Review by dholmes-ora — APPROVED** [review]
Hi Gerard,

Thanks for making the changes. I had hoped to completely isolate this to BSD/x86 (with Apple ifdefs) but it seems we are a bit inconsistent in platform management with VM_version code: we have os_cpu .cpp files but only cpu header files! Can't quite see the logic in that. :)

David

**Pull request closed without merging**